tests/target/issue-4257.rs

```diff
@@ -0,0 +1,18 @@
+#![feature(generic_associated_types)]
+#![allow(incomplete_features)]
+
+trait Trait<T> {
+    type Type<'a>
+    where
+        T: 'a;
+    fn foo(x: &T) -> Self::Type<'_>;
+}
+impl<T> Trait<T> for () {
+    type Type<'a>
+    where
+        T: 'a,
+    = &'a T;
```

Comment:
This is the correct behaviour right? Here's another similar example:
https://github.com/rust-lang/rustfmt/blob/e81c393663daab11238944cddfbd97541630f178/tests/target/issue-4159.rs#L15-L17